### Additional steps

In [1]:
from google.colab import drive

drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [2]:
from pathlib import Path
import pandas as pd
import gc
from sklearn.preprocessing import LabelEncoder
import pickle
import bz2
import matplotlib.pyplot as plt
import numpy as np

In [4]:
data_path = Path('./gdrive/MyDrive/STDS').resolve()
!ls {data_path}

09_prepared.pkl.bz2  10_prepared.pkl.bz2  11_prepared.pkl.bz2
09_stats.pkl.bz2     10_stats.pkl.bz2	  suspect.pkl.bz2


In [ ]:
with bz2.open(data_path / '09_prepared.pkl.bz2', 'rb') as f:
    df = pickle.load(f)
df.head()

In [6]:
df.drop(['id_card', 'id_doc'], axis=1, inplace=True)
df.head()

date  id_order  id_tov  ...  is_green  id_card_int  id_doc_int
0 2020-09-01 17:01:25         0      52  ...     False      1538855     1283228
1 2020-09-01 17:35:36         0      52  ...     False       267307     8873113
2 2020-09-01 19:19:21         0      52  ...     False       610220    12712899
3 2020-09-01 21:53:34         0      52  ...     False       441497    21535283
4 2020-09-01 18:42:31         0      61  ...     False      1065358      642341

[5 rows x 9 columns]

In [7]:
with bz2.open(data_path / '09_stats.pkl.bz2', 'wb') as f:
    pickle.dump(df, f, protocol=4)

In [4]:
with bz2.open(data_path / '10_prepared.pkl.bz2', 'rb') as f:
    df_2 = pickle.load(f)
df_2.head()

date  ... id_doc_int
0 2020-10-01 20:22:55  ...    2035558
1 2020-10-01 20:54:12  ...   13037327
2 2020-10-01 21:25:59  ...    1430590
3 2020-10-01 13:31:01  ...    3504660
4 2020-10-01 16:32:06  ...   11024438

[5 rows x 11 columns]

In [7]:
with bz2.open(data_path / '10_stats.pkl.bz2', 'wb') as f:
    pickle.dump(df_2, f, protocol=4)

In [9]:
with bz2.open(data_path / '10_stats.pkl.bz2', 'rb') as f:
    df_2 = pickle.load(f)
df_2.head()

date  ... id_doc_int
0 2020-10-01 20:22:55  ...    2035558
1 2020-10-01 20:54:12  ...   13037327
2 2020-10-01 21:25:59  ...    1430590
3 2020-10-01 13:31:01  ...    3504660
4 2020-10-01 16:32:06  ...   11024438

[5 rows x 11 columns]

In [5]:
with bz2.open(data_path / '11_prepared.pkl.bz2', 'rb') as f:
    df_3 = pickle.load(f)
df_3.head()

date  ... id_doc_int
0 2020-11-01 15:29:01  ...    3121751
1 2020-11-01 12:41:10  ...   22047017
2 2020-11-01 13:17:09  ...    6419825
3 2020-11-01 12:31:31  ...   10284541
4 2020-11-01 13:06:15  ...   10424967

[5 rows x 11 columns]

In [6]:
with bz2.open(data_path / '11_stats.pkl.bz2', 'wb') as f:
    pickle.dump(df_3, f, protocol=4)

In [7]:
with bz2.open(data_path / '11_stats.pkl.bz2', 'rb') as f:
    df_3 = pickle.load(f)
df_3.head()

date  ... id_doc_int
0 2020-11-01 15:29:01  ...    3121751
1 2020-11-01 12:41:10  ...   22047017
2 2020-11-01 13:17:09  ...    6419825
3 2020-11-01 12:31:31  ...   10284541
4 2020-11-01 13:06:15  ...   10424967

[5 rows x 11 columns]

### Statistics

In [8]:
from pathlib import Path
import pandas as pd
import gc
from sklearn.preprocessing import LabelEncoder
import pickle
import bz2

data_path = Path('./gdrive/MyDrive/STDS').resolve()
!ls {data_path}

09_prepared.pkl.bz2  09_stats.pkl.bz2  10_prepared.pkl.bz2  11_prepared.pkl.bz2


**The first month statistics**

In [6]:
with bz2.open(data_path / '09_stats.pkl.bz2', 'rb') as f:
    df = pickle.load(f)
df.head()

date  id_order  id_tov  ...  is_green  id_card_int  id_doc_int
0 2020-09-01 17:01:25         0      52  ...     False      1538855     1283228
1 2020-09-01 17:35:36         0      52  ...     False       267307     8873113
2 2020-09-01 19:19:21         0      52  ...     False       610220    12712899
3 2020-09-01 21:53:34         0      52  ...     False       441497    21535283
4 2020-09-01 18:42:31         0      61  ...     False      1065358      642341

[5 rows x 9 columns]

In [9]:
df.describe()

id_order        id_tov  ...   id_card_int    id_doc_int
count  4.210086e+07  4.210086e+07  ...  4.210086e+07  4.210086e+07
mean   7.765869e+05  1.763064e+04  ...  8.446032e+05  1.170384e+07
std    2.032896e+06  1.103226e+04  ...  5.388947e+05  6.760102e+06
min    0.000000e+00  4.600000e+01  ...  0.000000e+00  2.000000e+00
25%    0.000000e+00  1.452600e+04  ...  3.650470e+05  5.848724e+06
50%    0.000000e+00  1.943500e+04  ...  8.150010e+05  1.170182e+07
75%    0.000000e+00  2.455800e+04  ...  1.288708e+06  1.755596e+07
max    6.673056e+06  3.866000e+04  ...  1.944074e+06  2.341787e+07

[8 rows x 7 columns]

In [10]:
df['day'] = df['date'].dt.date
df.head()

date  id_order  id_tov  ...  id_card_int  id_doc_int         day
0 2020-09-01 17:01:25         0      52  ...      1538855     1283228  2020-09-01
1 2020-09-01 17:35:36         0      52  ...       267307     8873113  2020-09-01
2 2020-09-01 19:19:21         0      52  ...       610220    12712899  2020-09-01
3 2020-09-01 21:53:34         0      52  ...       441497    21535283  2020-09-01
4 2020-09-01 18:42:31         0      61  ...      1065358      642341  2020-09-01

[5 rows x 10 columns]

The first step I made to find outlier was grouping all orders by customers and look at the resulting numbers. The sorted dataframe shows that the largest number of orders is made by the user with id = 883386

In [14]:
df.groupby(['id_card_int']).size().reset_index(name="counts").sort_values(by="counts")

id_card_int  counts
41129          53184       1
654758        886496       1
230626        302034       1
1122584      1545446       1
1328580      1862246       1
...              ...     ...
163425        213588     675
389897        518424     694
910003       1243870     707
426570        568930     753
652503        883386  189812

[1358089 rows x 2 columns]

In [22]:
suspicious = df[df.id_card_int==883386]
suspicious.count()

date           189812
id_order       189812
id_tov         189812
id_kontr       189812
quantity       189812
sum            189812
is_green       189812
id_card_int    189812
id_doc_int     189812
day            189812
dtype: int64

In [25]:
suspicious[suspicious.id_order==0].count()

date           189812
id_order       189812
id_tov         189812
id_kontr       189812
quantity       189812
sum            189812
is_green       189812
id_card_int    189812
id_doc_int     189812
day            189812
dtype: int64

The above stats shows that all of his 189812 orders were made offline.

The next step was made to make sure that the customer with id_card = 883386 has some anomalies. I have calculated the overall amount that he spend and the difference between him and the second one is enormous.

In [30]:
df.groupby(['id_card_int']).sum().sort_values(by="sum")

id_order      id_tov  ...  is_green     id_doc_int
id_card_int                          ...                         
1871302               0       21435  ...         0       21829155
527413                0      179951  ...         5       86521061
1843017               0      234032  ...         0       23359304
405371                0      162821  ...         4       58061750
365257                0         647  ...         0        2932231
...                 ...         ...  ...       ...            ...
1447941      1846224652    11934022  ...         3     7513709846
1755035        19554147      279221  ...         2      205354227
1243870               0    14823178  ...         5     7581114411
568930       2234948749    17156402  ...         8     5952086098
883386                0  4163789442  ...       355  2226507914953

[1358089 rows x 7 columns]

The statistics for other months proves above ideas. Everything points to the exact same customer.

**October**

In [11]:
df_2.groupby(['id_card_int']).size().reset_index(name="counts").sort_values(by="counts")

id_card_int  counts
681980       928324       1
759204      1037262       1
71786         94324       1
759210      1037269       1
759215      1037274       1
...             ...     ...
225556       297027     730
331358       440354     769
399283       533540     772
317603       421742     881
649740       883386  204271

[1363961 rows x 2 columns]

**November**

In [8]:
df_3.groupby(['id_card_int']).size().reset_index(name="counts").sort_values(by="counts")

id_card_int  counts
1100156      1569326       1
558437        772791       1
954998       1350833       1
1279817      1855550       1
558430        772781       1
...              ...     ...
733034       1025552     732
365881        497966     737
1105409      1577248     789
1271558      1841799     790
634961        883386  228012

[1335358 rows x 2 columns]

In [14]:
suspect = pd.DataFrame(data=['883386'])
with bz2.open(data_path / 'suspect.pkl.bz2', 'wb') as f:
  pickle.dump(suspect, f, protocol=4)